### PR TITLE
Fix fast image border animation

### PR DESCRIPTION
**Pull request by hannojg**
### The issue

The `FastImageBorderRadiusAnimator#shouldAnimateProperty` returned false in cases where the `FastImage` component is a child of another component that has a border radius. Example for such a use case:

```jsx
<View style={{ borderRadius: 30, overflow: "hidden" }}>
   <FastImage {...} />
</View>
```

In this case the animation would look like this. You can see that the borders will "make a jump" once the SET is finished:

https://user-images.githubusercontent.com/16821682/118120763-aebbd380-b3f0-11eb-879f-1348b11721a0.mp4

### The fix

I wrote a function that checks recursively:
* If the view is a ReactView -> try to get border radius from this react view
* If the view is a FastImageView -> try to get border radius from this FastImage view
* If it founds a border radius it returns

https://user-images.githubusercontent.com/16821682/118122398-183ce180-b3f3-11eb-8df8-e0f0202a2a3f.mp4

### Open questions

The recursive function fails when it reaches the root with the following error
```
java.lang.ClassCastException: android.view.ViewRootImpl cannot be cast to android.view.ViewGroup
```
currently, I try/catch this. Maybe there is a better way to check this (like `view is Overlay`) ?

### CODE REVIEW

**Comment by hannojg** [comment]
Hmm, I think we don't need any of this anymore:

https://github.com/wix/react-native-navigation/blob/06eaa5f0620b7f3b512473c888962e123c38a2e4/lib/android/app/src/main/java/com/reactnativenavigation/views/element/animators/FastImageBorderRadiusAnimator%2Butilities.kt#L45-L61

This check earlier is enough, it will also gets the View's border radius that wraps the RNFastImage view. Not sure if this code has been build for any specific reason like that @guyca ?

File: lib/android/app/src/main/java/com/reactnativenavigation/views/element/animators/FastImageBorderRadiusAnimator+utilities.kt [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+ *
+ * <View style={{ borderRadius: 30, overflow: "hidden" }}>
+ *     <FastImage {...} />
+ * </View>
+ */
+fun FastImageBorderRadiusAnimator.getInheritedBorderRadius(v: View): Float {
+    val borderRadius = getBorderRadius(v)
+    if (borderRadius > 0f) {
+        return borderRadius
+    }
```
Comment (guyca) [inline lib/android/app/src/main/java/com/reactnativenavigation/views/element/animators/FastImageBorderRadiusAnimator+utilities.kt:26]:
```suggestion
    if (borderRadius > 0f) return borderRadius
```

File: lib/android/app/src/main/java/com/reactnativenavigation/views/element/animators/FastImageBorderRadiusAnimator+utilities.kt [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    return when(val parentView = getParent(v)) {
+        null -> 0f
+        else -> getInheritedBorderRadius(parentView)
+    }
+}
+
+private fun getBorderRadius(v: View): Float {
+    // gets the border radius from a react view
```
Comment (guyca) [inline lib/android/app/src/main/java/com/reactnativenavigation/views/element/animators/FastImageBorderRadiusAnimator+utilities.kt]:
```suggestion
```

File: lib/android/app/src/main/java/com/reactnativenavigation/views/element/animators/FastImageBorderRadiusAnimator+utilities.kt [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    if (v is ReactViewGroup && v.borderRadius > 0f) {
+        return v.borderRadius
+    }
+
+    // gets the border radius for a FastImage view
```
Comment (guyca) [inline lib/android/app/src/main/java/com/reactnativenavigation/views/element/animators/FastImageBorderRadiusAnimator+utilities.kt]:
```suggestion
```

File: lib/android/app/src/main/java/com/reactnativenavigation/views/element/animators/FastImageBorderRadiusAnimator+utilities.kt [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    return getBorderRadiusFastImage(v);
+}
+
+private fun getBorderRadiusFastImage(v: View): Float {
+    val parentView = getParent(v)
+
+    if (parentView == null || v is OverlayLayout) {
+        return 0f
+    }
```
Comment (guyca) [inline lib/android/app/src/main/java/com/reactnativenavigation/views/element/animators/FastImageBorderRadiusAnimator+utilities.kt]:
```suggestion
    if (parentView == null || v is OverlayLayout) return 0f
```

File: lib/android/app/src/main/java/com/reactnativenavigation/views/element/animators/FastImageBorderRadiusAnimator+utilities.kt [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+    val parentIsUsedOnlyToDrawBorderRadiusOverImage = parentView.childCount <= 1 || parentView.children.contains(v)
```
Comment (guyca) [inline lib/android/app/src/main/java/com/reactnativenavigation/views/element/animators/FastImageBorderRadiusAnimator+utilities.kt]: @hannojg Isn't `parentView.children.contains(v)` always true? 🤔

File: lib/android/app/src/main/java/com/reactnativenavigation/views/element/animators/FastImageBorderRadiusAnimator+utilities.kt [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+    return 0f;
+}
+
+private fun getParent(view: View): ViewGroup? = try {
+    when(view.parent) {
+        null -> null
+        else -> ViewTags.get<ViewGroup>(view, R.id.original_parent, view.parent as ViewGroup)
+    }
+// TODO: can we handle this better?: java.lang.ClassCastException: android.view.ViewRootImpl cannot be cast to android.view.ViewGroup
```
Comment (guyca) [inline lib/android/app/src/main/java/com/reactnativenavigation/views/element/animators/FastImageBorderRadiusAnimator+utilities.kt:49]: @hannojg Yes. Basically, we should return from the recursion earlier and not traverse up the hierarchy all the way to system views. I'll add the relevant changes as suggestions.

File: lib/android/app/src/main/java/com/reactnativenavigation/views/element/animators/FastImageBorderRadiusAnimator+utilities.kt [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    if (v is ReactViewGroup && v.borderRadius > 0f && v.overflow == "hidden") {
+        return v.borderRadius
+    }
+
+    return 0f;
+}
+
+private fun getParent(view: View): ViewGroup? = try {
+    when(view.parent) {
+        null -> null
```
Comment (guyca) [inline lib/android/app/src/main/java/com/reactnativenavigation/views/element/animators/FastImageBorderRadiusAnimator+utilities.kt:46]:
```suggestion
        null, is ReactView, is OverlayLayout -> null
```

This should prevent us from getting a ClassCastException. You can safely remove the try-catch.

File: lib/android/app/src/main/java/com/reactnativenavigation/views/element/animators/FastImageBorderRadiusAnimator.kt [hunk 8]
```diff
@@ -11,23 +11,31 @@ import androidx.core.animation.doOnEnd
...
-            Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP &&
-            fromChild !is ReactImageView
-                    && toChild !is ReactImageView
-                    && (getBorderRadius(from) != 0f || getBorderRadius(to) != 0f)
+    override fun shouldAnimateProperty(fromChild: ImageView, toChild: ImageView): Boolean {
+        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP &&
+                fromChild !is ReactImageView
+                && toChild !is ReactImageView
+                && (getInheritedBorderRadius(from) != 0f
+                || getInheritedBorderRadius(to) != 0f)
```
Comment (guyca) [inline lib/android/app/src/main/java/com/reactnativenavigation/views/element/animators/FastImageBorderRadiusAnimator.kt:31]:
```suggestion
                && (getInheritedBorderRadius(from) != 0f || getInheritedBorderRadius(to) != 0f)
```

**Review by guyca — CHANGES_REQUESTED** [review]

File: lib/android/app/src/main/java/com/reactnativenavigation/views/element/animators/FastImageBorderRadiusAnimator+utilities.kt [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * <View style={{ borderRadius: 30, overflow: "hidden" }}>
+ *     <FastImage {...} />
+ * </View>
+ */
+fun FastImageBorderRadiusAnimator.getInheritedBorderRadius(v: View): Float {
+    val borderRadius = getBorderRadius(v)
+    if (borderRadius > 0f) return borderRadius
+
+    if (v is ReactView || v is OverlayLayout) {
```
Comment (guyca) [inline lib/android/app/src/main/java/com/reactnativenavigation/views/element/animators/FastImageBorderRadiusAnimator+utilities.kt:28]: This check should be done first thing in the method as both ReactView and OverlayLayout are RNN internal layouts. Even if they had a border-radius, we don't want to use it in our calculations.

**Review by guyca — CHANGES_REQUESTED** [review]

**Review by guyca — APPROVED** [review]

**Comment by guyca** [comment]
The FastImageBorderRadiusAnimator#shouldAnimateProperty returned false in cases where the FastImage component is a child of another component that has a border radius.

**Pull request merged**